crates/wasm-encoder/src/core/code.rs

```diff
@@ -2728,6 +2728,79 @@ impl Encode for Instruction<'_> {
     }
 }
 
+/// A constant expression.
+///
+/// Usable in contexts such as offsets or initializers.
+#[derive(Debug)]
+pub struct ConstExpr {
+    bytes: Vec<u8>,
+}
+
+impl ConstExpr {
+    /// Create a new empty constant expression builder.
+    pub fn new() -> Self {
+        Self { bytes: Vec::new() }
+    }
+
+    /// Add raw bytes to this constant expression.
+    pub fn raw(&mut self, bytes: impl IntoIterator<Item=u8>) -> &mut Self {
+        self.bytes.extend(bytes);
+        self
+    }
+
+    fn insn(&mut self, insn: Instruction) -> &mut Self {
```

Comment:
This seems reasonable to me to also make `pub` in case anyone wants to put custom instructions/etc for one reason or another (although I like the concrete methods as well too)

Reply:
Yeah, my goal here was to ensure the builder only really allows const operations, and `raw` is already there as a backdoor for people to use.

Reply:
I left this as private for the time being. It does seem like expanding this type is going to be inevitable, so I feel like finding use-cases first makes sense.